homeassistant/components/greeneye_monitor/__init__.py

```diff
@@ -117,7 +117,9 @@
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the GreenEye Monitor component."""
     monitors = greeneye.Monitors()
-    hass.data[DATA_GREENEYE_MONITOR] = monitors
+    hass.data[DOMAIN] = {
+        DATA_MONITORS: monitors,
+    }
```

Comment:
Fair enough. What I'm planning to put in there is a `StorageCollection` to hold the state of some configuration entities that I'm adding as part of moving from YAML to UI config. However, all of that work is blocked behind #59871 clearing the PR queue, and I figured getting this little bit in now would reduce the unrelated clutter in the eventual configuration entities PR, making that one also easier to review when it arrives. This seemed like it would have been a reasonable way for me to have structured the data in the first place, even if the dict only had one entry, because the constant name for the key makes it clearer what the data is, and the dict leaves room for future expansion. I can hold off on this until actually adding the additional data.

Reply:
`StorageCollection` is meant for integrations that are configured for both YAML and via the UI. However, this integration communicates with a device or service and thus no YAML configuration is allowed.

Therefore the `StorageCollection` makes no sense IMHO?

Reply:
> Where should configuration entities store their state if it's not readable from the device itself?

Entities should always originate/control data points from the device itself (so-called "virtual devices" are not allowed in Home Assistant). We have a RestoreEntity base class that can help restore state from a previous run, which is generally used when working with device controls for which we have to assume its state (as no feedback is provided).

I'm not entirely sure about the context in this case.

Reply:
In this case (and I'm happy to move this conversation to an issue or some other place if you would prefer), the GreenEye Monitor has some pulse counter inputs on it. You can connect these to any sensor that emits pulses--water meters, gas meters, electric meters, for example. The GreenEye Monitor itself does not know what is attached to it; it just reports pulses.

So for Home Assistant integration, we need a way to map those pulses to an appropriate unit (L, gal, m^3, kWh, etc.) for whatever is producing them. Today that is done in the YAML config. 

My first attempt at a UI config flow put it in the options flow, but that produced a really awkward UI (select a monitor, select a pulse counter, specify how we should convert pulses to units). It also produced really awkward code--because the options flow has no way of accessing the entities themselves, whenever a new monitor connected I would edit the config entry and add default values to the options dict for all of the pulse counters. Then if a pulse counter was later deleted, I would edit the config entry again and clean those up.

I thought configuration entities would provide nicer UX (since they could appear on the device page for the pulse counter) and simpler code, but if those would be considered "virtual devices" I won't pursue that path.

Reply:
> So for Home Assistant integration, we need a way to map those pulses to an appropriate unit (L, gal, m^3, kWh, etc.) for whatever is producing them.

That is a hard concept for Home Assistant at this point, as we really do not have the appropriate construct to support those things actually 

> I thought configuration entities would provide nicer UX (since they could appear on the device page for the pulse counter) and simpler code, but if those would be considered "virtual devices" I won't pursue that path.

I might do that, but it isn't allowed.

In general, we want devices to represent their data as much "as is" as possible.